File: js/test/userTests.js
var chai = require('chai')
  , expect = chai.expect
  , should = chai.should();

var sinon = require('sinon');

describe('Regular User', function () {


  it('should have a method to list all movies', function () {

    // create a stub for the service
    var service = {
      getAllMovies: function(callback){
        callback(null, [{title:'Alien'}, {title:'Alien2'}]);
      }
    };
    // inject stub in user constructor
    var user = require('../src/user.js')(service, "Dominique", false);

    user.getAllMovies(function(err,data){
      expect(err).to.equal(null);
      expect(data).to.not.equal(null);
      expect(data.length).to.equal(2);
      expect(data[0].title).to.equal('Alien');
      expect(data[1].title).to.equal('Alien2');
    });
  });

  it('should have a method to list all movies with Mock service', function(){
    // create a mock
    var movieService = require('../src/movieService');
    var mockService = sinon.mock(movieService);
    // define the mock's expected behavior
    mockService.expects("getAllMovies").once();
    //inject into user
    var user = require('../src/user.js')(mockService.object, "Dominique", false);

    user.getAllMovies(function(err, data){
      // checks that the mock behaved as expected: should not output "Real service was called"
      mock.verify();
    });
  });

  it('should have a method to rent a movie', function () {
  });
  it('should have a method to list current rentals', function () {
  });
});

File: java/src/test/java/UserTests.java
import org.json.JSONObject;
import org.junit.Assert;
import org.junit.Ignore;
import org.junit.Test;

import java.io.IOException;

import static org.mockito.Mockito.*;

public class UserTests {

    @Test
    public void Any_User_Can_List_All_Movies() throws IOException {
        // this is a stub
        IMovieService stubService = new StubService();
        User user = new User(stubService , "Dom", false);
        JSONObject actual = user.getAllMovies();
        JSONObject expected = new JSONObject("{result:[{id:1, title:'Alien'}]}");
        // test that we receive the expected output
        Assert.assertEquals(expected.toString(), actual.toString());
    }

    @Test
    public void Any_User_Can_List_Current_Rentals() throws IOException {
        // this is a mock, using the Mockito framework
        IMovieService mockService = mock(IMovieService.class);
        when(mockService.getAllMovies()).thenReturn(new JSONObject("{result:'doesnt matter'}"));
        User user = new User(mockService , "John", false);
        user.getAllMovies();
        // test that the service method has been called exactly once
        verify(mockService).getAllMovies();
    }

    @Test
    @Ignore
    public void Any_User_Can_Rent_A_Movie() {
        // add code here
    }

    @Test
    @Ignore
    public void Regular_User_Cannot_Add_A_Movie() {
        // add code here
    }
}
